File: remez.py
import numpy as np
from numpy.fft import fft
import matplotlib.pyplot as plt


def find(vet, comp, maior):
    res = []
    if maior:
        for i in range(0, len(vet)):
            if vet[i] >= comp:
                res.append(i)
    else:
        for i in range(0, len(vet)):
            if vet[i] <= comp:
                res.append(i)

    return res


def defineK(R, f):
    vet = np.linspace(0, len(f)-1, int(R))
    aux = []
    for i in range(0, len(vet)):
        aux.append(f[round(vet[i])])
    return aux


def defineS(R):
    s = []
    for i in range(1, int(R+1)):
        s.append(((-1)**i))
    return s


def defineX(m, k, w, s, W, D):
    # Transforma os vetores em matrizes
    matrixm = m.reshape(1, len(m))
    matrixw = w[k].reshape(len(k), 1)
    # Multiplica as matrizes
    matrixwm = matrixw*matrixm
    # Calcula a matriz de cos
    coswm = np.cos(matrixwm)
    Wnp = np.array(W)

    # Calcula matriz que deve ser concat
    matrixsW = s/Wnp[k]

    # Concatena matrizes
    mF = np.zeros((coswm.shape[0], coswm.shape[1]+1))
    mF[:, :-1] = coswm
    mF[:, -1:] = matrixsW.reshape(coswm.shape[0], 1)

    # Calcula solução linear para mF \ D[k]
    matrixD = np.array(D)
    x = np.linalg.solve(mF, matrixD[k])
    return x


def defineA(x, M):
    a = []
    for i in range(0, int(M)+1):
        a.append(x[i])
    return a


def defineH(a, M):
    vet = []
    for i in range(int(M), 0, -1):
        vet.append(a[i])
    vet.append(2*a[0])
    for i in range(1, int(M)+1):
        vet.append(a[i])
    vet = np.array(vet)
    vet = vet/2
    return vet


def defineW(w):
    vet = []
    for i in range(0, len(w)):
        vet.append(Ks1*(w[i] <= ws1) + Kp*((w[i] >= wp1)
                                           and (w[i] <= wp2)) + Ks2*(w[i] >= ws2))
    return vet


def defineD(w):
    vet = []
    for i in range(0, len(w)):
        vet.append(int((wp1 <= w[i]) and (w[i] <= wp2)))
    return vet


def defineNovoK(vet1, vet2):
    res = []
    for i in range(len(vet1)):
        res.append(vet1[i])
    for i in range(len(vet2)):
        res.append(vet2[i])
    res = sorted(res)

    return res


def defineErroK(A, D, W, novok):
    errok = []
    for i in range(len(novok)):
        errok.append((A[novok[i]] - D[novok[i]])*W[novok[i]])
    
    return errok


def defineV(errok, delta, stop):
    res = []
    for i in range(0, len(errok)):
        if abs(errok[i]) >= abs(delta-stop):
            res.append(True)
    return res

def firamp(h, tipo, L):
    N = len(h)
    H = fft(h, 2*L)
    H = H[0:L+1]
    w = np.arange(0, L+1)*np.pi/L
    M = (N-1)/2

    if tipo == 1 or tipo == 2:
        H = np.exp(M*1j*w)*H
    else:
        H = -1j*np.exp(M*1j*w)*H

    A = H.real
    return A


def localMax(vet):
    n = len(vet)
    vcomp1 = []
    vcomp2 = []
    res = []

    vcomp1.append(vet[0]-1)
    for i in range(0, n-1):
        vcomp1.append(vet[i])

    for i in range(1, n):
        vcomp2.append(vet[i])
    vcomp2.append(vet[n-1]-1)

    for i in range(0, n):
        if vet[i] > vcomp1[i] and vet[i] > vcomp2[i]:
            res.append(i)

    return res


# Função que implementa o algoritmo de remez
# para construção de filtros FIR
def remez(N, D, W):
    L = len(W) - 1
    stop = 10**(-8)
    M = (N-1)/2
    R = M + 2

    # Inicializando conjunto de referências
    f = find(W, stop, True)
    k = defineK(R, f)
    w = np.arange(0, L+1)*np.pi/L
    m = np.arange(0, M+1)
    s = defineS(R)

    while 1:
        # Resolvendo problema de interpolação
        x = defineX(m, k, w, s, W, D)
        a = defineA(x, M)
        delta = x[int(M)+1]
        h = defineH(a, M)
        A = firamp(h, 1, L)
        erro = (A - D)*W  # Erro do peso

        # Atualizando conjunto de referências
        novok = defineNovoK(localMax(erro), localMax(-erro))
        errok = defineErroK(A, D, W, novok)

        # Remove frequencias em que o erro do peso é menor que delta
        v = defineV(errok, delta, stop)
        print(v)
        
        break

    return 0  # h, delta


# Script de teste
N = 103
ws1 = 0.20*np.pi
wp1 = 0.25*np.pi
wp2 = 0.60*np.pi
ws2 = 0.70*np.pi
Ks1 = 10
Kp = 1
Ks2 = 1
wo = (ws1+wp1)/2
w1 = (wp1+ws2)/2
L = 4000
w = np.arange(0, L+1)*np.pi/L
W = defineW(w)
D = defineD(w)
h = remez(N, D, W)

File: Matlab/remez.m
function [h,delta] = remez(N,D,W)
% h = remez(N,D,W)
% Design de filtros tipo 1
%
% h : Resposta ao impulso de tamanho N
% N : Tamanho do filtro (Impar)
% D : Resposta Ideal (grid uniforme)
% W : Função de peso (grid uniforme)
% O tamanho de N deve ser igual ao de D

W = W(:);
D = D(:);
L = length(W)-1;
stop = 1e-8; % Numero pequeno definido para critério de parada
M = (N-1)/2;
R = M + 2; % R = Tamanho do conjunto de referencia

% Inicializa conjunto de referência (Espaçado igualmente aproximadamente onde W>0)
f = find(W>stop);
k = f(round(linspace(1,length(f),R)));
w = (0:L)'*pi/L;
m = 0:M;
s = (-1).^(1:R)'; % Sinais
while 1
% Resolvendo problema de interpolação
x = [cos(w(k)*m), s./W(k)] \ D(k);
a = x(1:M+1); 
delta = x(M+2); 
h = [a(M+1:-1:2); 2*a(1); a(2:M+1)]/2;
A = firamp(h,1,L)';
erro = (A-D).*W; % Erro do peso

% Atualizando conjunto de referências
novok = sort([localMax(erro); localMax(-erro)]);
errok = (A(novok)-D(novok)).*W(novok);

% Remove frequencias em que o erro do peso é menor que delta
v = abs(errok) >= (abs(delta)-stop);
novok = novok(v);
errok = errok(v);

% Garante a propriedade de alternação
v = gpalt(errok);
novok = novok(v);
errok = errok(v);

% Se o novo conjunto for muito grande, remove pontos até atingir tamanho
% adequado
while length(novok) > R
if abs(errok(1)) < abs(errok(length(novok)))
novok(1) = [];
else
novok(length(novok)) = [];
end
end

% Verifica se convergiu
if (max(errok)-abs(delta))/abs(delta) < stop
disp('Convergiu')
break
end
k = novok;
end
delta = abs(delta);
h = [a(M+1:-1:2); 2*a(1); a(2:M+1)]/2;
